paf-mvp-client-express/src/client-node.ts

```diff
@@ -67,417 +56,420 @@ export interface ClientNodeConfig extends Config {
-
-    const tld = getTopLevelDomain(hostName);
-
-    // Only allow calls from the same TLD+1, under HTTPS
-    const allowedOrigins: RegExp[] = [
-      new RegExp(
-        `^https:\\/\\/(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\\-]*[a-zA-Z0-9])\\.)*${escapeRegExp(tld)}(/?$|\\/.*$)`
-      ),
-    ];
-
-    const isValidOrigin = (origin: string) => allowedOrigins.findIndex((regexp: RegExp) => regexp.test(origin)) !== -1;
-
-    const checkOrigin = (endpoint: string) => (req: Request, res: Response, next: () => unknown) => {
-      const origin = req.header('origin');
-
-      if (isValidOrigin(origin)) {
-        next();
-      } else {
-        const error: ClientNodeError = {
-          type: ClientNodeErrorType.INVALID_ORIGIN,
-          details: `Origin is not allowed: ${origin}`,
-        };
-        logger.Error(endpoint, error);
-        res.status(400);
-        res.json(error);
-      }
-    };
-
-    const checkReferer = (endpoint: string) => (req: Request, res: Response, next: () => unknown) => {
-      const referer = req.header('referer');
-
-      if (isValidOrigin(referer)) {
-        next();
-      } else {
-        const error: ClientNodeError = {
-          type: ClientNodeErrorType.INVALID_REFERER,
-          details: `Referer is not allowed: ${referer}`,
-        };
-        logger.Error(endpoint, error);
-        res.status(400);
-        res.json(error);
-      }
-    };
+    this.client = new OperatorClient(operatorHost, hostName, currentPrivateKey, this.keyStore);
+    this.postIdsPrefsRequestBuilder = new PostIdsPrefsRequestBuilder(operatorHost, hostName, currentPrivateKey);
+    this.get3PCRequestBuilder = new Get3PCRequestBuilder(operatorHost);
+    this.getNewIdRequestBuilder = new GetNewIdRequestBuilder(operatorHost, hostName, currentPrivateKey);
+    this.deleteIdsPrefsRequestBuilder = new DeleteIdsPrefsRequestBuilder(operatorHost, hostName, currentPrivateKey);
 
-    const corsOptions: CorsOptions = {
-      origin: allowedOrigins,
-      optionsSuccessStatus: 200, // some legacy browsers (IE11, various SmartTVs) choke on 204
-      credentials: true,
-      allowedHeaders: ['Content-Type'],
-    };
-    // FIXME methods for each endpoint
+    this.originValidator = new OriginValidator(hostName, this.logger);
 
     // *****************************************************************************************************************
-    // ************************************************************************************************************ JSON
+    // ************************************************************************************************************ REST
     // *****************************************************************************************************************
+    this.app.expressApp.get(
+      jsonProxyEndpoints.read,
+      this.originValidator.cors,
+      this.originValidator.checkOrigin(jsonProxyEndpoints.read),
```

Comment:
Yes, but it accepts requests with **no** or empty `origin` header.
These requests (that would probably be triggered by a S2S call, not from a browser) will fail at the operator level (see https://github.com/prebid/addressability-framework/blob/main/mvp-spec/security-signatures.md#:~:text=S2S%20call%20to%20the%20OneKey%20client%20node%20with%20fake%20origin%20HTTP%20header). But it's better to refuse them here and avoid a failing call to the operator.
Since your question is very legit, I added a comment in the code.

Reply:
In this case, would it be valid that the header simply exists?

Reply:
Yes, it would probably be sufficient to test:
1. cors with either empty or valid origin header
2. check that header is not empty

However, isn't safer the way it is today?

Reply:
I agree.
I think I found a good compromise with my latest commit: I only check that the header is not empty, however I made this check part of the `cors` handler.
This way, it is impossible to have only 1 **_or_** 2.

Thanks for your insistence on this.

Reply:
(I need more tests, commit coming soon)

Reply:
Ok, it's good I made more tests.
In fact, I had a wrong interpretation of how CORS works:

The cors middleware is just a helper to set the `Access-Control-Allow-Origin` response header. In case the `Origin` request header is not valid, then _nothing particular happens server side_.

It's **the browser** that compares and decides to trigger a CORS error.

So, the real mandatory check is the "custom" one (`checkOrigin`), because it will return a `40x` error if the `Origin` request header is not valid, even in the case of a S2S.
So we could only keep the `checkOrigin` handler, but I think it's more standard to also set the cors middleware so that browser will show a CORS error in the console, if the origin value is not valid.

Reply:
In the end, I added a comment in code, left the two handlers, but optimized slightly by setting just a single regexp rather than an array of regexp.